### PR TITLE
[video_player] Update texture on seekTo

### PR DESCRIPTION
**Pull request by KevinTheGray**
## Description

@cbenhagen I basically copied your PR: https://github.com/flutter/plugins/pull/2397,  and opened a new one here with changes after the big 0.10.10 update.  It's still a very simple change.  ~~I see that they want XC tests, but I need some clarification on how that is expected to work.~~ 

~~@cyanglaz  I see the PR you reference on the other PR, (https://github.com/flutter/plugins/pull/2445) but I don't understand how to run those tests.  I commented on it with the issues I am seeing.  If I could understand how I am expected to install the mock dependency and run the test, I could add the tests to this PR.~~  Test added!

### What this fixes
Explicitly mark texture frame available on seekTo. This is necessary since flutter/engine@6a90b7f has changed the external texture data flow from a producer-consumer model. Now ios_external_texture_gl caches previous opengl texture, if no new frame are available, it does not execute the copyPixelBuffer method, just uses cached opengl texture to draw.

## Related Issues

Fixes flutter/flutter#43567

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by dthuering** [comment]
Hi @KevinTheGray 
can you resolve the merge conflicts? I am looking forward to having this issue fixed. :)

**Comment by KevinTheGray** [comment]
@dthuering thanks for the heads up! resolved it.

If you want, you can also give this PR a thumbs up and the original issue here a thumbs up: https://github.com/flutter/flutter/issues/43567

I believe that's part of how they prioritize which ones to review and merge in

**Comment by jameswasher** [comment]
I’m also eagerly awaiting this PR.

File: packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m
Comment (dannyvalentesonos) [inline packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:585]: I've used this trick, and it doesn't always work. Unfortunately, what can happen is when Flutter comes to request the frame, the iOS AVPlayer can say there's no frame available and this results in a black frame being rendered.

To fix it, I've actually had to unpause the display link and pause it again after 0.5 seconds. It won't pause again if a play was requested in the mean time, and I also handle getting a new seek request in the middle as well.

My solution works perfectly and never results in any black frames.

Comment (KevinTheGray) [reply packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:585]: @dannyvalentesonos Sorry for the incredibly late reply, been busy with work.  Can you create a small repo that shows your fix in action?

**Comment by jpiabrantes** [comment]
What's the status on this? This PR would be really important for my app :)

**Comment by stuartmorgan** [comment]
@KevinTheGray Does https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#setting-up-xcuitests and https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#xctests-ios address your question about running tests? If not, can you clarify what exactly you're looking for help with?

**Comment by KevinTheGray** [comment]
@stuartmorgan most likely! I'll give those a look and follow up if not.  Sorry, have just been keeping this branch up to date but haven't dug into the testing.  Hopefully I can have something soon.  Thanks

**Comment by KevinTheGray** [comment]
@stuartmorgan so I currently have two problems understanding how these tests are to be written.  Taking the `webview` package as an example 

1.) The tests aren't passing for me in `webview` when I run `pod lib lint --allow-warnings`
I changed to the iOS directory and get the following error when I run that command:
```
pod lib lint --allow-warnings

 -> webview_flutter (0.0.1)
    - ERROR | [iOS] xcodebuild: Returned an unsuccessful exit code. You can use `--verbose` for more information.
    - NOTE  | xcodebuild:  note: Using new build system
    - NOTE  | xcodebuild:  note: Building targets in parallel
    - NOTE  | xcodebuild:  note: Using codesigning identity override: -
    - NOTE  | xcodebuild:  note: Planning build
    - NOTE  | xcodebuild:  note: Constructing build description
    - NOTE  | [iOS] xcodebuild:  warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.0, but the range of supported deployment target versions is 9.0 to 14.2.99. (in target 'App' from project 'App')
    - NOTE  | [iOS] xcodebuild:  warning: Skipping code signing because the target does not have an Info.plist file and one is not being generated automatically. (in target 'App' from project 'App')
    - NOTE  | [iOS] xcodebuild:  warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.0, but the range of supported deployment target versions is 9.0 to 14.2.99. (in target 'Pods-App' from project 'Pods')
    - NOTE  | [iOS] xcodebuild:  ld: warning: ignoring file /Users/KG/Library/Developer/Xcode/DerivedData/App-grtmcjsybkgxiofopejgpjphnagi/Build/Products/Release-iphonesimulator/webview_flutter/webview_flutter.framework/webview_flutter, building for iOS Simulator-arm64 but attempting to link with file built for iOS Simulator-x86_64
    - NOTE  | [iOS] xcodebuild:  clang: error: linker command failed with exit code 1 (use -v to see invocation)
    - NOTE  | [iOS] xcodebuild:  ld: warning: ignoring file /Users/KG/Library/Developer/Xcode/DerivedData/App-grtmcjsybkgxiofopejgpjphnagi/Build/Products/Release-iphonesimulator/webview_flutter/webview_flutter.framework/webview_flutter, building for iOS Simulator-i386 but attempting to link with file built for iOS Simulator-x86_64
    - NOTE  | [iOS] xcodebuild:  ld: warning: ignoring file Flutter/Flutter.framework/Flutter, missing required architecture i386 in file Flutter/Flutter.framework/Flutter (3 slices)
    - NOTE  | xcodebuild:  warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.0, but the range of supported deployment target versions is 9.0 to 14.2.99. (in target 'Flutter' from project 'Pods')
    - NOTE  | xcodebuild:  warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.0, but the range of supported deployment target versions is 9.0 to 14.2.99. (in target 'webview_flutter' from project 'Pods')
    - NOTE  | [iOS] xcodebuild:  warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.0, but the range of supported deployment target versions is 9.0 to 14.2.99. (in target 'OCMock' from project 'Pods')
    - NOTE  | [iOS] xcodebuild:  warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 8.0, but the range of supported deployment target versions is 9.0 to 14.2.99. (in target 'webview_flutter-Unit-Tests' from project 'Pods')
```

2. I can't get the imports or autocomplete to work in the test files.  If I open the example project and open `FLTWKNavigationDelegateTests.m` or `FLTWebviewTests.m`, `@import Flutter;` fails saying that it can't be found.  Is there a specific way I'm supposed to open a file/project to get autocomplete to work?  

This is my flutter doctor:
```
[✓] Flutter (Channel master, 1.27.0-9.0.pre.126, on Mac OS X 10.15.4 19E287 darwin-x64, locale en-US)
    • Flutter version 1.27.0-9.0.pre.126 at /Users/KG/Developer/Flutter/flutter
    • Framework revision e81e0b850c (3 days ago), 2021-02-27 19:36:02 -0500
    • Engine revision 5c14c528cc
    • Dart version 2.13.0 (build 2.13.0-81.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/KG/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /Users/KG/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.2, Build version 12B45b
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage
        on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] IntelliJ IDEA Community Edition (version 2018.2.5)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 31.3.3
    • Dart plugin version 182.5215

[✓] VS Code (version 1.53.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0
```

File: packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m
Comment (ened) [reply packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:585]: @dannyvalentesonos do you have a sample or reference to your approach?

Comment (dannyvalentesonos) [reply packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:585]: I don't have a repo that only has this change unfortunately. I will try and find some time to get one together.

**Comment by ldelafuente** [comment]
Thanks guys for your work! Do you have any updates? It's been more than a year since this and other similar PRs where created. We urgently need this for production, otherwise we can't launch.

**Comment by cbenhagen** [comment]
@ldelafuente I wouldn't hold my breath on this. The regression was introduced in August 2019 already and PR's with a fix exist since December 2019. The team is slowly digging through the backlog of plugin PR's and the situation is getting better every month. If you need this today I would recommend to maintain your own fork of the plugin with the fix or just use the branch of this PR in a dependency override.

**Comment by stuartmorgan** [comment]
@KevinTheGray There's now more documentation about plugin tests at https://github.com/flutter/flutter/wiki/Plugin-Tests, and basic test harnesses have been backfilled in the iOS implementations of the plugins (see [here](https://github.com/flutter/plugins/blob/master/packages/video_player/video_player/example/ios/RunnerTests/VideoPlayerTests.m) for video_player) which should make adding new tests much easier.

**Comment by KevinTheGray** [comment]
@stuartmorgan Thanks! I'll merge ad take a look.  Hope to have something soon

File: packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m
Comment (KevinTheGray) [reply packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:600]: I'm going resolve this.  I've been using this fix for over a year now and the client hasn't had any complaints

**Comment by KevinTheGray** [comment]
@stuartmorgan  Sorry for the delay, but the test that was originally requested is now in. The requirements were as follows: 
```
We need some test, how about an XCTest that mocks the texture registry, invokes handleMethodCall with a "seekTo" and asserts that textureFrameAvailable was invoked?
```
I've mocked the registry and invoked `seekTo` directly on the pigeon Api (the plugin has changed a lot since then and no longer uses methodCalls).  Then I verified that textureFrameWasAvailable

**Comment by KevinTheGray** [comment]
@stuartmorgan I think this should be as you suggested now.  Let me know if I didn't get it right, happy to make any changes!

**Comment by machty** [comment]
I know we're still waiting on a few tweaks to this PR before it can be merged, but can some other people chime in to mention whether this PR fixes the seek behavior in their apps?

File: packages/video_player/video_player/CHANGELOG.md
Comment (stuartmorgan) [inline packages/video_player/video_player/CHANGELOG.md:3]: Nit: Updates (per new style guide for CHANGELOGs)

**Review by stuartmorgan — APPROVED** [review]
> I will make sure to review future updates here promptly.

😞 

LGTM! @bparrishMines or @cyanglaz for secondary review (per policy).

**Review by bparrishMines — APPROVED** [review]
LGTM

**Comment by ened** [comment]
LGTM from my side too. We have been using a version of this PR for quite a while in production. CC @guymclean

**Review by ened — APPROVED** [review]

**Pull request merged**
